src/librbd/ImageWatcher.cc

```diff
@@ -1061,7 +1077,6 @@ void ImageWatcher<I>::handle_rewatch(int r) {
 
   Context *unregister_watch_ctx = nullptr;
   {
-    RWLock::WLocker watch_locker(m_watch_lock);
```

Comment:
Nit: need to hold the lock

Reply:
Counter-nit: I do, I just moved it up earlier because it should already be held in the retry_rewatch branch ... ;)

Reply:
The reason why the lock was in the block was to avoid invoking the callback w/ the lock held.

Reply:
Ah, so, not about having to hold the lock but not holding the lock at a later point. 

My personal style would explicitly unlock the lock before calling the callback then: this makes  the intention more explicit and avoids more complicated duplication ... OK?

Reply:
It's your code (since this won't be merged upstream)

Reply:
Fair enough. :) 

(Although I was hoping we'd do the same dance with merging it into the Jewel branch even though it doesn't get any more releases.)

Reply:
Ok, I got out of lost puppy mode. I chose to make it a bit dirty (even though I did read up a bit on C++) and leverage scoping instead of calling unlock() directly. I've compiled this in our environment and am trying to cause some stress there to see whether we're good now.